### PR TITLE
[Compiler] Fix function invocation

### PR DESCRIPTION
**Pull request by SupunS**
Work towards https://github.com/onflow/cadence/issues/3804

## Description

Includes couple of fixes related to invocation compilation:
- Fix invoking a function-typed field. e.g. `foo.bar()`, where `bar` is a function-typed field, rather than a method of `foo`.
  - This should be invoked in the same way function pointers are being invoked.
  - i.e: Should use the `Invoke` instruction, and should not be invoked using `InvokeMethodStatic`/`InvokeMethodDynamic`.
- Change the expression evaluation order of invocation expressions to match that of interpreter.
  - Invoked expression must be evaluated first, and then the arguments.

An unrelated fix:
- Always transfer the elements in for-loop.

______

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: interpreter/misc_test.go [hunk 1]
```diff
@@ -39,316 +39,30 @@ import (
...
 	"github.com/onflow/cadence/sema"
 	"github.com/onflow/cadence/stdlib"
+	"github.com/onflow/cadence/test_utils"
 	. "github.com/onflow/cadence/test_utils/common_utils"
 	. "github.com/onflow/cadence/test_utils/interpreter_utils"
 	. "github.com/onflow/cadence/test_utils/runtime_utils"
 	. "github.com/onflow/cadence/test_utils/sema_utils"
 )
 
-type ParseCheckAndInterpretOptions struct {
```
Comment (SupunS) [inline interpreter/misc_test.go:48]: These functions were replicated to `github.com/onflow/cadence/test_utils` with a hope to remove it from here, in a previous PR. So remove the implementation form there, but keeps names as-is just to avoid having to refactor all the tests (by converting them to variables with the same name) .

File: interpreter/misc_test.go [hunk 2]
```diff
@@ -39,316 +39,30 @@ import (
 	"github.com/onflow/cadence/runtime"
 	"github.com/onflow/cadence/sema"
 	"github.com/onflow/cadence/stdlib"
+	"github.com/onflow/cadence/test_utils"
```
Comment (turbolent) [inline interpreter/misc_test.go:42]: Maybe make this a dot-import, like with the other test utils packages

**Review by turbolent — APPROVED** [review]
Great work! 👏

File: interpreter/misc_test.go (hunk 2 above)
Comment (SupunS) [reply interpreter/misc_test.go:42]: Not possible yet, because some duplicate types are also defined in this file (`misc_test.go`) as well. Would need to refactor all tests that uses them and then we can safely remove the duplicates from here. Will do that in a separate PR.
